### PR TITLE
ifelseternary

### PR DESCRIPTION
**Pull request by burner**
A check for patterns

```D
string f;
if(cond) {
   f = "Hello";
} else {
   f = "World";
}
```

to be replaced by 
```D
string f = cond ? "Hello" : "World";
```

### CODE REVIEW

**Comment by Basile-z** [comment]
Good job, there are quite a bunch of matches in [phobos](https://travis-ci.org/dlang-community/D-Scanner/jobs/517741244#L285) .

**Comment by WebFreak001** [comment]
I think this feature is great but right now it's way too eager.

https://github.com/dlang-community/D-Scanner/pull/758/files#diff-7c757b7c4d7a2962179c7846a5669b88R135
I don't know how to even rewrite this as nice ternary, this if-else is much more readable than anything I can think of.

Can the diagnostics message also suggest the replacement code? It would also make it much easier for text editors to offer automatic replace commands. Of course this would be even easier if dscanner had these fixes built-in like having the source range and a replacement text to insert. - like #351 

Also I think it would be better if it wasn't a warning, but some kind of new diagnostic like "suggestion" which editors wouldn't neccessarily mark as `"warning: your code is bad"`, but rather only show when you hover over the affected source range (or similar)

But I would love to have this feature because sometimes there are cases where a ternary would fit much better.

**Comment by burner** [comment]
> I don't know how to even rewrite this as nice ternary, this if-else is much more readable than anything I can think of.

```D
string person = "unknown";
string hello;
const personUnknown = person == "unknown";
person = personUnknown ? "text" : "bar;
hello = hello.empty && personUnknown ? "foo" : hello;
```

I don't think D-Scanner supports suggestions or automatic rewriting.
I'm open to suggestions, but I don't want to create a new subsystem or anything along that lines to get this in.
And you can always disable this test in your local config.

File: src/dscanner/analysis/ifelseternary.d [hunk 1]
```diff
@@ -0,0 +1,172 @@
+//          Copyright Brian Schott (Hackerpilot) 2014.
```
Comment (Hackerpilot) [inline src/dscanner/analysis/ifelseternary.d]: More time-traveling identity theft :trollface:

**Comment by Hackerpilot** [comment]
This case is not detected:

```d
void main(string[] args)
{
    int a;
    if (a == 100)
        a = to!int(args[1]);
    else
        a = 100;
}
```

**Comment by burner** [comment]
@Hackerpilot is there a clean way to get the left side of an assign expression?

**Comment by burner** [comment]
> This case is not detected:
> 
> ```d
> void main(string[] args)
> {
>     int a;
>     if (args.length > 1)
>         a = to!int(args[1]);
>     else
>         a = 100;
> }
> ```

I added a test with that code.

**Comment by Hackerpilot** [comment]
> @Hackerpilot is there a clean way to get the left side of an assign expression?

All of the AST classes have the line `mixin OpEquals;`, which generates an `==` overload that will recursively scan both nodes. This is what I use in the check for both branches of an `if` statement being the same.

You should be able to check that
1. Both branches of the `if` have a single statement.
2. Both statements are assign expressions.
3. The operators for both assign expressions are `=`
4. The left sides of both expressions are equal according to the `==` overload.

**Comment by Basile-z** [comment]
> 3. The operators for both assign expressions are =

The check would be valid for other assign operators as long as both are identical.

**Comment by burner** [comment]
that is not really what I'm looking for.
I'm looking for a way to get the identifiers of the left hand side of the AssignExpressions in the then and else branch.

**Comment by Basile-z** [comment]
if you want the text of the left side text you can use the ast formatter.

**Comment by burner** [comment]
hm, that seems like overkill. I have a implemented a solution in this PR though.

**Comment by burner** [comment]
so what is the way forward?

**Comment by Basile-z** [comment]
Well let's forget the operator for now so this is simple. Memorize the whole LHS expression, and compare (== as explained by HackerPilot)  the one of the if branch with one of the else branch to see if the target is the same. Then to display the LHS exp use the formatter, it's very simple too.

**Comment by Basile-z** [comment]
In the absence of willing to compare the LHS assignment, whatever is its complexity, in the same scope the semantic should be the same.

**Pull request closed without merging**